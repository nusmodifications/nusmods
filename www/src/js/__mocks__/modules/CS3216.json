{
  "ModuleCode": "CS3216",
  "ModuleTitle": "Software Product Engineering for Digital Markets",
  "AcadYear": "2017/2018",
  "Department": "Computer Science",
  "ModuleDescription": "In this module, students will practice software product engineering by working in small teams to develop well-tested, user-friendly, production-quality software for the real world. To support this goal, students work closely with users to understand their problems, gather their requirements, and obtain their feedback through a rapid, iterative, application design and development process. Students will also be exposed to practical issues for digital markets such as growing the user base of their application, deployment of the application on the Web or in the cloud system, and validating the UI design and UX of the application.",
  "ModuleCredit": "5",
  "Workload": "2-1-0-8-2",
  "Types": [
    "Module"
  ],
  "Prerequisite": "CS2103 or its equivalent or with special approval from instructor. Students will submit personal statements to apply for a place in the course instead of bidding through the CORS system.",
  "History": [
    {
      "Semester": 1,
      "Timetable": [
        {
          "ClassNo": "1",
          "LessonType": "Lecture",
          "WeekText": "Every Week",
          "DayText": "Monday",
          "StartTime": "1830",
          "EndTime": "2030",
          "Venue": "VCRm"
        }
      ],
      "LecturePeriods": [
        "Monday Evening"
      ]
    }
  ],
  "ParsedPrerequisite": "CS2103",
  "ModmavenTree": {
    "name": "CS3216",
    "children": {
      "name": "CS2103",
      "children": []
    }
  },
  "LockedModules": []
}
