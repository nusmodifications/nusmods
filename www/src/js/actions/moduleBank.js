// @flow
import type { AcadYear, Module, ModuleCode } from 'types/modules';
import type { TimetableConfig } from 'types/timetables';

import type { GetState } from 'types/redux';
import type { ModulesMap } from 'reducers/moduleBank';
import { size, get, flatMap, sortBy, zip } from 'lodash';
import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';

const MAX_MODULE_LIMIT: number = 100;

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

// Action to fetch modules
export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ?ModuleCode {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export const UPDATE_MODULE_TIMESTAMP = 'UPDATE_MODULE_TIMESTAMP';
export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export const REMOVE_LRU_MODULE = 'REMOVE_LRU_MODULE';
export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result) => {
        onFinally();
        return result;
      },
      (error) => {
        onFinally();
        throw error;
      },
    );
  };
}

// Action to fetch module from previous years
export const FETCH_ARCHIVE_MODULE: string = 'FETCH_ARCHIVE_MODULE';
export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Array<Module>>(config.archiveYears, modules));
}
