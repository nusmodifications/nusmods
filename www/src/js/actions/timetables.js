// @flow
import { each, flatMap } from 'lodash';
import localforage from 'localforage';

import type { ModuleLessonConfig, SemTimetableConfig } from 'types/timetables';
import type { FSA, GetState } from 'types/redux';
import type { ColorIndex, ColorMapping } from 'types/reducers';
import type { ClassNo, Lesson, LessonType, Module, ModuleCode, Semester } from 'types/modules';

import { fetchModule } from 'actions/moduleBank';
import { randomModuleLessonConfig, validateTimetableModules } from 'utils/timetables';
import { getModuleTimetable } from 'utils/modules';
import storage from 'storage';
import { V2_MIGRATION_KEY } from 'storage/keys';
import { MIGRATION_KEYS, parseQueryString } from 'storage/migrateTimetable';

export const ADD_MODULE: string = 'ADD_MODULE';
export function addModule(semester: Semester, moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) =>
    dispatch(fetchModule(moduleCode)).then(() => {
      const module: Module = getState().moduleBank.modules[moduleCode];
      const lessons = getModuleTimetable(module, semester);
      const moduleLessonConfig = randomModuleLessonConfig(lessons);

      return dispatch({
        type: ADD_MODULE,
        payload: {
          semester,
          moduleCode,
          moduleLessonConfig,
        },
      });
    });
}

export const REMOVE_MODULE: string = 'REMOVE_MODULE';
export function removeModule(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: REMOVE_MODULE,
    payload: {
      semester,
      moduleCode,
    },
  };
}

export const MODIFY_LESSON: string = 'MODIFY_LESSON';
export function modifyLesson(activeLesson: Lesson): FSA {
  return {
    type: MODIFY_LESSON,
    payload: {
      activeLesson,
    },
  };
}

export const CHANGE_LESSON: string = 'CHANGE_LESSON';
export function setLesson(
  semester: Semester,
  moduleCode: ModuleCode,
  lessonType: LessonType,
  classNo: ClassNo,
): FSA {
  return {
    type: CHANGE_LESSON,
    payload: {
      semester,
      moduleCode,
      lessonType,
      classNo,
    },
  };
}

export function changeLesson(semester: Semester, lesson: Lesson): FSA {
  return setLesson(semester, lesson.ModuleCode, lesson.LessonType, lesson.ClassNo);
}

export const CANCEL_MODIFY_LESSON: string = 'CANCEL_MODIFY_LESSON';
export function cancelModifyLesson(): FSA {
  return {
    type: CANCEL_MODIFY_LESSON,
    payload: null,
  };
}

export const SET_TIMETABLE = 'SET_TIMETABLE';
export function setTimetable(
  semester: Semester,
  timetable?: SemTimetableConfig,
  colors?: ColorMapping,
) {
  return (dispatch: Function, getState: GetState) => {
    let validatedTimetable = timetable;
    if (timetable) {
      const moduleCodes = getState().moduleBank.moduleCodes;
      [validatedTimetable] = validateTimetableModules(timetable, moduleCodes);
    }

    return dispatch({
      type: SET_TIMETABLE,
      payload: { semester, timetable: validatedTimetable, colors },
    });
  };
}

export function fillTimetableBlanks(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    const { timetables, moduleBank } = getState();

    // Extract the timetable and the modules for the semester
    const timetable = timetables.lessons[semester];
    if (!timetable) return;

    // Check that all lessons for each module is filled, if they are not, use the
    // randomly generated config to fill them in
    each(timetable, (lessonConfig: ModuleLessonConfig, moduleCode: ModuleCode) => {
      const module = moduleBank.modules[moduleCode];
      if (!module) return;

      const lessons = getModuleTimetable(module, semester);
      const randomLessonConfig = randomModuleLessonConfig(lessons);

      each(randomLessonConfig, (classNo: ClassNo, lessonType: LessonType) => {
        if (!lessonConfig[lessonType]) {
          dispatch(setLesson(semester, moduleCode, lessonType, classNo));
        }
      });
    });
  };
}

export function fetchTimetableModules(timetables: SemTimetableConfig[]) {
  return (dispatch: Function) => {
    const moduleCodes = new Set(flatMap(timetables, Object.keys));
    return Promise.all(
      Array.from(moduleCodes).map((moduleCode) => dispatch(fetchModule(moduleCode))),
    );
  };
}

export function migrateTimetable() {
  return (dispatch: Function, getState: GetState): Promise<*> => {
    if (storage.getItem(V2_MIGRATION_KEY)) {
      return Promise.resolve();
    }

    const promises = MIGRATION_KEYS.map(([semester, key]) =>
      localforage.getItem(key).then((queryString) => {
        // Do nothing if there's no data
        if (!queryString) return null;

        const timetable = parseQueryString(queryString);
        const [validTimetable] = validateTimetableModules(
          timetable,
          getState().moduleBank.moduleCodes,
        );

        dispatch(setTimetable(semester, validTimetable));
        return dispatch(fetchTimetableModules([timetable])).then(() =>
          dispatch(fillTimetableBlanks(semester)),
        );
      }),
    );

    return Promise.all(promises).then(() => storage.setItem(V2_MIGRATION_KEY, true));
  };
}

export const SELECT_MODULE_COLOR: string = 'SELECT_MODULE_COLOR';
export function selectModuleColor(
  semester: Semester,
  moduleCode: ModuleCode,
  colorIndex: ColorIndex,
): FSA {
  return {
    type: SELECT_MODULE_COLOR,
    payload: {
      semester,
      moduleCode,
      colorIndex,
    },
  };
}

export const HIDE_LESSON_IN_TIMETABLE: string = 'HIDE_LESSON_IN_TIMETABLE';
export function hideLessonInTimetable(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: HIDE_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}

export const SHOW_LESSON_IN_TIMETABLE: string = 'SHOW_LESSON_IN_TIMETABLE';
export function showLessonInTimetable(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: SHOW_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}
