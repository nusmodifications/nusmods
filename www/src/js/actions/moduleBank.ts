import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';
import { flatMap, get, size, sortBy, zip } from 'lodash';
import { ModulesMap } from 'types/reducers';
import { AcadYear, Module, ModuleCode } from 'types/modules';

import { GetState } from 'types/redux';
import { TimetableConfig } from 'types/timetables';
import {
  FETCH_ARCHIVE_MODULE,
  FETCH_MODULE,
  FETCH_MODULE_LIST,
  REMOVE_LRU_MODULE,
  UPDATE_MODULE_TIMESTAMP,
} from './constants';

const MAX_MODULE_LIMIT = 100;

export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ModuleCode | null {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove: ModuleCode[] = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy<ModuleCode>(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result: any) => {
        onFinally();
        return result;
      },
      (error: any) => {
        onFinally();
        throw error;
      },
    );
  };
}

export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Module[]>(config.archiveYears, modules));
}
