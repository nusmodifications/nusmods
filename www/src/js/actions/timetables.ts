import { each, flatMap } from 'lodash';

import { Lesson, ColorIndex, ModuleLessonConfig, SemTimetableConfig } from 'types/timetables';
import { FSA, GetState } from 'types/redux';
import { ColorMapping } from 'types/reducers';
import { ClassNo, LessonType, Module, ModuleCode, Semester } from 'types/modules';

import { fetchModule } from 'actions/moduleBank';
import { openNotification } from 'actions/app';
import { getModuleCondensed } from 'selectors/moduleBank';
import {
  randomModuleLessonConfig,
  validateModuleLessons,
  validateTimetableModules,
} from 'utils/timetables';
import { getModuleTimetable } from 'utils/modules';

export const ADD_MODULE = 'ADD_MODULE';
export function addModule(semester: Semester, moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) =>
    dispatch(fetchModule(moduleCode)).then(() => {
      const module: Module = getState().moduleBank.modules[moduleCode];

      if (!module) {
        dispatch(
          openNotification(`Cannot load ${moduleCode}`, {
            action: {
              text: 'Retry',
              handler: () => dispatch(addModule(semester, moduleCode)),
            },
          }),
        );

        return;
      }

      const lessons = getModuleTimetable(module, semester);
      const moduleLessonConfig = randomModuleLessonConfig(lessons);

      dispatch({
        type: ADD_MODULE,
        payload: {
          semester,
          moduleCode,
          moduleLessonConfig,
        },
      });
    });
}

export const REMOVE_MODULE = 'REMOVE_MODULE';
export function removeModule(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: REMOVE_MODULE,
    payload: {
      semester,
      moduleCode,
    },
  };
}

export const MODIFY_LESSON = 'MODIFY_LESSON';
export function modifyLesson(activeLesson: Lesson): FSA {
  return {
    type: MODIFY_LESSON,
    payload: {
      activeLesson,
    },
  };
}

export const CHANGE_LESSON = 'CHANGE_LESSON';
export function setLesson(
  semester: Semester,
  moduleCode: ModuleCode,
  lessonType: LessonType,
  classNo: ClassNo,
): FSA {
  return {
    type: CHANGE_LESSON,
    payload: {
      semester,
      moduleCode,
      lessonType,
      classNo,
    },
  };
}

export function changeLesson(semester: Semester, lesson: Lesson): FSA {
  return setLesson(semester, lesson.moduleCode, lesson.lessonType, lesson.classNo);
}

export const SET_LESSON_CONFIG = 'SET_LESSON_CONFIG';
export function setLessonConfig(
  semester: Semester,
  moduleCode: ModuleCode,
  lessonConfig: ModuleLessonConfig,
): FSA {
  return {
    type: SET_LESSON_CONFIG,
    payload: {
      semester,
      moduleCode,
      lessonConfig,
    },
  };
}

export const CANCEL_MODIFY_LESSON = 'CANCEL_MODIFY_LESSON';
export function cancelModifyLesson(): FSA {
  return {
    type: CANCEL_MODIFY_LESSON,
    payload: null,
  };
}

export const SET_TIMETABLE = 'SET_TIMETABLE';
export function setTimetable(
  semester: Semester,
  timetable?: SemTimetableConfig,
  colors?: ColorMapping,
) {
  return (dispatch: Function, getState: GetState) => {
    let validatedTimetable = timetable;
    if (timetable) {
      [validatedTimetable] = validateTimetableModules(timetable, getState().moduleBank.moduleCodes);
    }

    return dispatch({
      type: SET_TIMETABLE,
      payload: { semester, timetable: validatedTimetable, colors },
    });
  };
}

export function validateTimetable(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    const { timetables, moduleBank } = getState();

    // Extract the timetable and the modules for the semester
    const timetable = timetables.lessons[semester];
    if (!timetable) return;

    // Check that all lessons for each module are valid. If they are not, we update it
    // such that they are
    each(timetable, (lessonConfig: ModuleLessonConfig, moduleCode: ModuleCode) => {
      const module = moduleBank.modules[moduleCode];
      if (!module) return;

      const [validatedLessonConfig, changedLessonTypes] = validateModuleLessons(
        semester,
        lessonConfig,
        module,
      );

      if (changedLessonTypes.length) {
        dispatch(setLessonConfig(semester, moduleCode, validatedLessonConfig));
      }
    });
  };
}

export function fetchTimetableModules(timetables: SemTimetableConfig[]) {
  return (dispatch: Function, getState: GetState) => {
    const moduleCodes = new Set(flatMap(timetables, Object.keys));
    const validateModule = getModuleCondensed(getState().moduleBank);

    return Promise.all(
      Array.from(moduleCodes)
        .filter(validateModule)
        .map((moduleCode) => dispatch(fetchModule(moduleCode))),
    );
  };
}

export const SELECT_MODULE_COLOR = 'SELECT_MODULE_COLOR';
export function selectModuleColor(
  semester: Semester,
  moduleCode: ModuleCode,
  colorIndex: ColorIndex,
): FSA {
  return {
    type: SELECT_MODULE_COLOR,
    payload: {
      semester,
      moduleCode,
      colorIndex,
    },
  };
}

export const HIDE_LESSON_IN_TIMETABLE = 'HIDE_LESSON_IN_TIMETABLE';
export function hideLessonInTimetable(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: HIDE_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}

export const SHOW_LESSON_IN_TIMETABLE = 'SHOW_LESSON_IN_TIMETABLE';
export function showLessonInTimetable(semester: Semester, moduleCode: ModuleCode): FSA {
  return {
    type: SHOW_LESSON_IN_TIMETABLE,
    payload: { moduleCode, semester },
  };
}
