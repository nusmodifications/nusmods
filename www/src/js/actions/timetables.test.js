// @flow
import localforage from 'localforage';
import type { ModuleCode, Semester, Lesson } from 'types/modules';
import type { SemTimetableConfig } from 'types/timetables';

import lessons from '__mocks__/lessons-array.json';

/** @var {Module} */
import CS1010S from '__mocks__/modules/CS1010S.json';
/** @var {Module} */
import CS3216 from '__mocks__/modules/CS3216.json';

import runThunk from 'test-utils/runThunk';
import { V2_MIGRATION_KEY } from 'storage/keys';
import * as actions from './timetables';
import { FETCH_MODULE } from './moduleBank';

// Workaround for Jest not being able to recognize implicitly mocked modules
const storage = (require('storage'): any); // eslint-disable-line global-require

jest.mock('localforage', () => ({ getItem: jest.fn() }));
jest.mock('storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

// see: https://github.com/reactjs/redux/blob/master/docs/recipes/WritingTests.md#example-1
// TODO: write addModule test with nock and mockStore.
test('addModule should create an action to add a module', () => {
  const moduleCode: ModuleCode = 'CS1010';
  const semester: Semester = 1;

  const value: Function = actions.addModule(semester, moduleCode);
  // TODO
  expect(typeof value === 'function').toBe(true);
});

test('removeLesson should return information to remove module', () => {
  const semester: Semester = 1;
  const moduleCode: ModuleCode = 'CS1010';
  expect(actions.removeModule(semester, moduleCode)).toMatchSnapshot();
});

test('modifyLesson should return lesson payload', () => {
  const activeLesson: Lesson = lessons[0];
  expect(actions.modifyLesson(activeLesson)).toMatchSnapshot();
});

test('changeLesson should return updated information to change lesson', () => {
  const semester: Semester = 1;
  const lesson: Lesson = lessons[1];
  expect(actions.changeLesson(semester, lesson)).toMatchSnapshot();
});

test('cancelModifyLesson should not have payload', () => {
  expect(actions.cancelModifyLesson()).toMatchSnapshot();
});

test('select module color should dispatch a select of module color', () => {
  const semester: Semester = 1;
  expect(actions.selectModuleColor(semester, 'CS1010S', 0)).toMatchSnapshot();
  expect(actions.selectModuleColor(semester, 'CS3216', 1)).toMatchSnapshot();
});

describe('migrateTimetable()', () => {
  const action = actions.migrateTimetable();
  const getState = () => ({
    moduleBank: { moduleCodes: { CS5331: {} }, modules: {} },
    timetables: { lessons: {} },
  });
  const makeDispatch = () => jest.fn().mockReturnValue(Promise.resolve());

  afterEach(() => {
    storage.setItem.mockReset();
  });

  test('not migrate if the timetable has already been migrated', async () => {
    const dispatch = makeDispatch();
    storage.getItem.mockReturnValue(true);

    await runThunk(action, dispatch, getState);

    expect(localforage.getItem).not.toHaveBeenCalled();
  });

  test('not migrate if old data is not present', async () => {
    const dispatch = makeDispatch();

    storage.getItem.mockReturnValue();
    localforage.getItem.mockReturnValue(Promise.resolve());
    await runThunk(action, dispatch, getState);

    expect(dispatch).not.toHaveBeenCalled();
    expect(storage.setItem).toHaveBeenCalledTimes(1);
    expect(storage.setItem).toHaveBeenCalledWith(V2_MIGRATION_KEY, true);
  });

  test('to migrate timetable', async () => {
    const dispatch = makeDispatch();
    storage.getItem.mockReturnValue();
    // Just mock one semester
    localforage.getItem.mockImplementation((key) => {
      if (key.includes('sem1')) return Promise.resolve('CS5331=');
      return Promise.resolve();
    });

    await runThunk(action, dispatch, getState);
    expect(dispatch).toHaveBeenCalledTimes(2);

    const [[firstAction], [secondAction]] = dispatch.mock.calls;
    expect(firstAction).toHaveProperty('type', actions.SET_TIMETABLE);
    expect(secondAction).toHaveProperty('type', FETCH_MODULE);

    expect(storage.setItem).toHaveBeenCalledTimes(1);
    expect(storage.setItem).toHaveBeenCalledWith(V2_MIGRATION_KEY, true);
  });

  test('to exclude invalid modules from timetable', async () => {
    const dispatch = makeDispatch();
    storage.getItem.mockReturnValue();
    // Just mock one semester
    localforage.getItem.mockImplementation((key) => {
      if (key.includes('sem1')) return Promise.resolve('CS5331=&DEADBEEF=');
      return Promise.resolve();
    });

    await runThunk(action, dispatch, getState);
    expect(dispatch).toHaveBeenCalledTimes(2);

    expect(storage.setItem).toHaveBeenCalledTimes(1);
    expect(storage.setItem).toHaveBeenCalledWith(V2_MIGRATION_KEY, true);
  });
});

describe('fillTimetableBlanks', () => {
  const moduleBank = { modules: { CS1010S, CS3216 } };
  const timetablesState = (semester: Semester, timetable: SemTimetableConfig) => ({
    lessons: { [semester]: timetable },
  });
  const semester = 1;
  const action = actions.fillTimetableBlanks(semester);

  test('do nothing if timetable is already full', () => {
    const timetable = {
      CS1010S: {
        Lecture: '1',
        Tutorial: '1',
        Recitation: '1',
      },
    };

    const state: any = { timetables: timetablesState(semester, timetable), moduleBank };
    const dispatch = jest.fn();
    action(dispatch, () => state);

    expect(dispatch).not.toHaveBeenCalled();
  });

  test('fill missing lessons with randomly generated modules', () => {
    const timetable = {
      CS1010S: {
        Lecture: '1',
        Tutorial: '1',
      },
      CS3216: {},
    };
    const state: any = { timetables: timetablesState(semester, timetable), moduleBank };
    const dispatch = jest.fn();

    action(dispatch, () => state);

    expect(dispatch).toHaveBeenCalledTimes(2);

    const [[firstAction], [secondAction]] = dispatch.mock.calls;
    expect(firstAction).toMatchObject({
      type: actions.CHANGE_LESSON,
      payload: {
        semester,
        moduleCode: 'CS1010S',
        lessonType: 'Recitation',
        classNo: expect.any(String),
      },
    });

    expect(secondAction).toMatchObject({
      type: actions.CHANGE_LESSON,
      payload: {
        semester,
        moduleCode: 'CS3216',
        lessonType: 'Lecture',
        classNo: '1',
      },
    });
  });
});

describe('hide/show timetable modules', () => {
  const semester: Semester = 1;

  test('should dispatch a module code for hiding', () => {
    const moduleCode: ModuleCode = 'CS1010';
    expect(actions.hideLessonInTimetable(semester, moduleCode)).toMatchSnapshot();
  });

  test('should dispatch a module code for showing', () => {
    const moduleCode: ModuleCode = 'CS1020';
    expect(actions.showLessonInTimetable(semester, moduleCode)).toMatchSnapshot();
  });
});
