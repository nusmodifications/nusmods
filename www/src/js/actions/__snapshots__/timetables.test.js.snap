// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelModifyLesson should not have payload 1`] = `
Object {
  "payload": null,
  "type": "CANCEL_MODIFY_LESSON",
}
`;

exports[`changeLesson should return updated information to change lesson 1`] = `
Object {
  "payload": Object {
    "classNo": "1",
    "lessonType": "Recitation",
    "moduleCode": "CS1010S",
    "semester": 1,
  },
  "type": "CHANGE_LESSON",
}
`;

exports[`hide/show timetable modules should dispatch a module code for hiding 1`] = `
Object {
  "payload": Object {
    "moduleCode": "CS1010",
    "semester": 1,
  },
  "type": "HIDE_LESSON_IN_TIMETABLE",
}
`;

exports[`hide/show timetable modules should dispatch a module code for showing 1`] = `
Object {
  "payload": Object {
    "moduleCode": "CS1020",
    "semester": 1,
  },
  "type": "SHOW_LESSON_IN_TIMETABLE",
}
`;

exports[`modifyLesson should return lesson payload 1`] = `
Object {
  "payload": Object {
    "activeLesson": Object {
      "ClassNo": "1",
      "DayText": "Wednesday",
      "EndTime": "1200",
      "LessonType": "Lecture",
      "ModuleCode": "CS1010S",
      "StartTime": "1000",
      "Venue": "LT26",
      "WeekText": "Every Week",
    },
  },
  "type": "MODIFY_LESSON",
}
`;

exports[`removeLesson should return information to remove module 1`] = `
Object {
  "payload": Object {
    "moduleCode": "CS1010",
    "semester": 1,
  },
  "type": "REMOVE_MODULE",
}
`;

exports[`select module color should dispatch a select of module color 1`] = `
Object {
  "payload": Object {
    "colorIndex": 0,
    "moduleCode": "CS1010S",
    "semester": 1,
  },
  "type": "SELECT_MODULE_COLOR",
}
`;

exports[`select module color should dispatch a select of module color 2`] = `
Object {
  "payload": Object {
    "colorIndex": 1,
    "moduleCode": "CS3216",
    "semester": 1,
  },
  "type": "SELECT_MODULE_COLOR",
}
`;
